GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,216 @@
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeStringWriter)} is not started");
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.isRunning = false;
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public void Dispose()
+        {
+            if (!this.isDisposed)
+            {
+                this.Stop();
+
+                this.pipeClient?.Dispose();
+                this.pipeClient = null;
+
+                this.writerThread = null;
+            }
+
+            this.isDisposed = true;
+        }
+
+        private void RaiseStateChanged(QueuedPipeStringWriterState newState, Exception ex)
+        {
+            int prevState = Interlocked.Exchange(ref this.state, (int)newState);
+
+            if (prevState != (int)newState)
+            {
+                this.StateChanged?.Invoke(this, new QueuedPipeStringWriterStateEventArgs(newState,  ex));
+            }
+        }
+
+        private void ThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue
+            do
+            {
+                try
+                {
+                    // Take() is a blocking call that will return a message when there is one available
+                    // or will throw an InvalidOperationException when the queue has been marked as completed.
+                    string message = this.queue.Take();
+                    if (message != null)
+                    {
+                        this.WriteMessage(message);
+                    }
+                }
+                catch (InvalidOperationException)
+                {
+                    // The queue is both empty *and* no longer accepting new messages
+                    Debug.Assert(this.queue.IsCompleted, "Queue should be marked as completed");
+                }
+            }
+            while (this.queue.Count > 0 || this.isRunning);
```

Comment:
Should this be `while (!this.queue.IsCompleted)` and remove the `isRunning` member? `TryAdd` will still throw when the  `CompleteAdding` is called so can we get away with using that as the flag?

Reply:
True, but we've no way now to prevent `TryEnqueue` when we've not yet called `Start` which sets-up and starts the draining thread. Perhaps that would be a desirable feature? (the ability to populate a queue of messages before dispatching them)